third_party/terraform/resources/resource_compute_router_nat.go

```diff
@@ -0,0 +1,392 @@
+
+	if v, ok := d.GetOk("nat_ip_allocate_option"); ok {
+		nat.NatIpAllocateOption = v.(string)
+	}
+
+	if v, ok := d.GetOk("nat_ips"); ok {
+		nat.NatIps = convertStringArr(v.(*schema.Set).List())
+	}
+
+	if v, ok := d.GetOk("source_subnetwork_ip_ranges_to_nat"); ok {
+		nat.SourceSubnetworkIpRangesToNat = v.(string)
+	}
+
+	if v, ok := d.GetOk("subnetwork"); ok {
+		nat.Subnetworks = expandSubnetworks(v.([]map[string]interface{}))
+	}
+
+	if v, ok := d.GetOk("min_ports_per_vm"); ok {
+		nat.MinPortsPerVm = int64(v.(int))
+	}
+
+	if v, ok := d.GetOk("udp_idle_timeout_sec"); ok {
+		nat.UdpIdleTimeoutSec = int64(v.(int))
+	}
+
+	if v, ok := d.GetOk("icmp_idle_timeout_sec"); ok {
+		nat.IcmpIdleTimeoutSec = int64(v.(int))
+	}
+
+	if v, ok := d.GetOk("tcp_established_idle_timeout_sec"); ok {
+		nat.TcpEstablishedIdleTimeoutSec = int64(v.(int))
+	}
+
+	if v, ok := d.GetOk("tcp_transitory_idle_timeout_sec"); ok {
+		nat.TcpTransitoryIdleTimeoutSec = int64(v.(int))
+	}
+
+	log.Printf("[INFO] Adding nat %s", natName)
+	nats = append(nats, nat)
+	patchRouter := &computeBeta.Router{
+		Nats: nats,
+	}
+
+	log.Printf("[DEBUG] Updating router %s/%s with nats: %+v", region, routerName, nats)
+	op, err := routersService.Patch(project, region, router.Name, patchRouter).Do()
+	if err != nil {
+		return fmt.Errorf("Error patching router %s/%s: %s", region, routerName, err)
+	}
+	d.SetId(fmt.Sprintf("%s/%s/%s", region, routerName, natName))
+	err = computeBetaOperationWaitTime(config.clientCompute, op, project, "Patching router", 4)
+	if err != nil {
+		d.SetId("")
+		return fmt.Errorf("Error waiting to patch router %s/%s: %s", region, routerName, err)
+	}
+
+	return resourceComputeRouterNatRead(d, meta)
+}
+
+func resourceComputeRouterNatRead(d *schema.ResourceData, meta interface{}) error {
+
+	config := meta.(*Config)
+
+	region, err := getRegion(d, config)
+	if err != nil {
+		return err
+	}
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	routerName := d.Get("router").(string)
+	natName := d.Get("name").(string)
+
+	routersService := config.clientComputeBeta.Routers
+	router, err := routersService.Get(project, region, routerName).Do()
+	if err != nil {
+		if gerr, ok := err.(*googleapi.Error); ok && gerr.Code == 404 {
+			log.Printf("[WARN] Removing router nat %s because its router %s/%s is gone", natName, region, routerName)
+			d.SetId("")
+
+			return nil
+		}
+
+		return fmt.Errorf("Error Reading router %s/%s: %s", region, routerName, err)
+	}
+
+	for _, nat := range router.Nats {
+
+		if nat.Name == natName {
+			d.SetId(fmt.Sprintf("%s/%s/%s", region, routerName, natName))
+			d.Set("nat_ip_allocate_option", nat.NatIpAllocateOption)
+			d.Set("nat_ips", schema.NewSet(schema.HashString, convertStringArrToInterface(nat.NatIps)))
+			d.Set("source_subnetwork_ip_ranges_to_nat", nat.SourceSubnetworkIpRangesToNat)
+			d.Set("subnetwork", flattenRouterNatSubnetworkToNatBeta(nat.Subnetworks))
```

Comment:
For any field that's a nested object (I think just this one), can you do:
```
if err := d.Set(...); err != nil {
  return fmt.Errorf("Error reading Router Nat: %s", err)
}
```

Reply:
Done.